src/fluree/db/ledger/transact/json.clj

```diff
@@ -0,0 +1,648 @@
+                    (<? (txfunction/execute object _id pred-info tx-state))
+                    object)]
+      (cond
+        (nil? object*) (throw (ex-info (str "Multi-cardinality values cannot be null/nil: " (pred-info :name))
+                                       {:status 400 :error :db/invalid-tx}))
+
+        (= :ref type) (cond
+                        (tempid/TempId? object*) object*
+                        (string? object*) (tempid/use object* tx-state)
+                        (int? object*) (<? (resolve-ident-strict object* tx-state))
+                        (util/pred-ident? object*) (<? (resolve-ident-strict object* tx-state)))
+
+        (= :tag type) (<? (tags/resolve object* pred-info tx-state))
+
+        :else (conform-object-value object* type)))))
+
+
+(defn resolve-object
+  "Resolves object into its final state so can be used for consistent comparisons with existing data."
+  [object _id pred-info tx-state]
+  (let [multi? (pred-info :multi)]
+    (if (nil? object)
+      (async/go :delete)                                    ;; delete any existing object
+      (cond-> object
+              (not multi?) (resolve-object-item _id pred-info tx-state)
+              multi? (->> (mapv #(resolve-object-item % _id pred-info tx-state))
+                          async/merge
+                          (async/into []))
+              (pred-info :unique) (resolve-unique _id pred-info tx-state)))))
+
+
+(defn add-singleton-flake
+  "Adds new-flake assuming not a duplicate. A retract-flake (if not nil)
+  is a matching flake in the existing db (i.e. a single-cardinality
+  flake must retract an existing single-cardinality value if it already
+  exists).
+
+  Performs some logic to determine if the new flake should get added at
+  all (i.e. if retract-flake is identical to the new flake)."
+  [flakes ^Flake new-flake ^Flake retract-flake pred-info tx-state]
+  (cond
+    ;; no retraction flake, always add
+    (nil? retract-flake)
+    (cond-> (conj flakes new-flake)
+            (pred-info :spec) (tx-validate/queue-pred-spec new-flake pred-info tx-state)
+            (pred-info :txSpec) (tx-validate/queue-predicate-tx-spec [new-flake] pred-info tx-state))
+
+    ;; new and retraction flake are identical
+    (= (.-o new-flake) (.-o retract-flake))
+    (if (pred-info :retractDuplicates)
+      ;; no need for predicate spec, new flake is exactly same as old
+      (cond-> (conj flakes new-flake retract-flake)
+              (pred-info :txSpec) (tx-validate/queue-predicate-tx-spec [new-flake retract-flake] pred-info tx-state))
+      ;; don't add new or retract flake
+      flakes)
+
+    ;; new and retraction flakes are different
+    :else
+    (cond-> (conj flakes new-flake retract-flake)
+            (pred-info :spec) (tx-validate/queue-pred-spec new-flake pred-info tx-state)
+            (pred-info :txSpec) (tx-validate/queue-predicate-tx-spec [new-flake retract-flake] pred-info tx-state))))
+
+
+(defn generate-statements
```

Comment:
Maybe not specifically for this pr and for discussion instead as this is more subjective, but I find functions this big with lots of nested loops really hard to follow. 

It's helpful for my understanding when I come across code I haven't seen before to find one main function that does (almost) nothing but call a bunch of helpers one after another, and possibly stringing those function calls together with `->` or `->>`. 

This provides a few advantages in my opinion:
- the helper function names can serve as free, lightweight documentation if they are named well. I can figure out how everything works at a high level by glancing at the "main" function, looking at the function names, and looking at the order they're called in 
- I can dive into the explicit details by focusing on individual helper functions at a time without unrelated code to distract me because that code is in different function bodies. 
- Each helper function provides even more documentation opportunities because they all have individual docstrings. 

Again, I don't have a hard blocker for this specific pr, but what do you think about that pattern?

Reply:
Completely agree, and where I could I did that throughout this PR... this is the one sore spot. I think I have this one broken up in ~half in the work on feature/prefix, so I don't expect it to last long in this exact state.

While this ballooned a little bit over the refactor, part of the reason this wasn't broken up farther (it is about half the size of the original) is there are a lot of conditional async chans. By wrapping them many functions, it results in creating additional channels that are not needed. This is the core processor, so a large transaction this could mean 100K+ extra async channels that were not needed. I'm concerned about wrapping that many values around channels that never needed to be wrapped, and how that might impact performance in this very performance-critical part of the code.